brushfire-core/src/main/scala/com/stripe/brushfire/Tree.scala

```diff
@@ -31,6 +32,107 @@ case class Tree[K, V, T](root: Node[K, V, T]) {
     }
   }
 
+  /**
+   * Prune a tree to minimize validation error.
+   * Replaces a split with a leaf when it would have a lower error than the child leaves.
+   *
+   * @param validationData
+   * @param start
+   * @param voter
+   * @param error
+   * @param weightMonoid
+   * @param errorOrdering
+   * @tparam P
+   * @tparam E
+   * @return
+   */
+  def prune[P, E](validationData: Map[Int, T], start: Option[Node[K, V, T]] = None, voter: Voter[T, P], error: Error[T, P, E])(implicit weightMonoid: Monoid[T], errorOrdering: Ordering[E]): Tree[K, V, T] = {
+    val begin: Node[K, V, T] = start.getOrElse(this.root) // Use root node as default root.
+    Tree(pruneNode(validationData, begin, voter, error))
+      .growByLeafIndex { i => Nil } //this will renumber all the leaves
+  }
+
+  /**
+   * Prune a tree to minimize validation error, starting from given root node.
+   *
+   * This method recursively traverses up the tree from the root, branching on splits, until it finds leaves, then
+   * goes back down the tree combining leaves when such a combination would reduce validation error.
+   *
+   * @param validationData
+   * @param start
+   * @param voter
+   * @param error
+   * @param weightMonoid
+   * @param errorOrdering
+   * @tparam P
+   * @tparam E
+   * @return
+   */
+  def pruneNode[P, E](validationData: Map[Int, T], start: Node[K, V, T], voter: Voter[T, P], error: Error[T, P, E])(implicit weightMonoid: Monoid[T], errorOrdering: Ordering[E]): Node[K, V, T] = {
+    val vData = mutable.Map() ++ validationData // Create a mutable Map to allow merging of validation data as a side effect of the pruneLevel method.
+    start match {
+      case leaf: LeafNode[K, V, T] => leaf // Bounce at the bottom and start back up the tree.
+      case split: SplitNode[K, V, T] => {
+        val newNode = SplitNode(children = split.children.map { case (k, p, n) => (k, p, pruneNode(validationData, n, voter, error)) })
+        newNode.children.exists {
+          case (k, v, s: SplitNode[K, V, T]) => true // If we have any splits as children, we can't prune.
+          case _ => false // If all children are leaves, we can prune.
+        } match {
+          case true => newNode
+          case false => pruneLevel(newNode, newNode.children.asInstanceOf[Seq[(K, Predicate[V], LeafNode[K, V, T])]], vData, voter, error)
```

Comment:
this asInstanceOf is really ugly. I think you'd be better to do something like this:

````scala
val childLeaves = newNode.children.collect{case (k,v,s: LeafNode[K,V,T]) => (k,v,s)}
if(childLeaves.size == newNode.children.size)
  pruneLevel(newNode, childLeaves, vData, voter, error)
else
  newNode
````